# Retrieving and cleaning the dataframe

In [49]:
import pandas as pd
import numpy as np

url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
df = pd.read_html(url)[0]

In [50]:
df.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned                        NaN
1         M2A      Not assigned                        NaN
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [62]:
# Removing all the Not assigned rows from Borough column.
df = df[~df['Borough'].isin(['Not assigned'])]

In [75]:
df.head(10)

Postal Code           Borough                                 Neighborhood
2          M3A        North York                                    Parkwoods
3          M4A        North York                             Victoria Village
4          M5A  Downtown Toronto                    Regent Park, Harbourfront
5          M6A        North York             Lawrence Manor, Lawrence Heights
6          M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
8          M9A         Etobicoke                             Islington Avenue
9          M1B       Scarborough                               Malvern, Rouge
11         M3B        North York                                    Don Mills
12         M4B         East York              Parkview Hill, Woodbine Gardens
13         M5B  Downtown Toronto                     Garden District, Ryerson

In [76]:
post = df['Postal Code'].value_counts()
post.values.tolist()
type(post)

pandas.core.series.Series

In [88]:
#Checking if there is any duplicated postal code
#for i in post:
#    if i != 1:
#        print('Algo')
#    else:
#        print('No Postal Code is duplicated')
    

#### <font color=red>All the 'Not Assigned' boroughs had 'Nan' Neighborhoods, therefore, they were all excluded.</font> 
#### <font color=red>Afterwards, no duplicated postal codes were found</font> 

In [82]:
df.shape

(103, 3)

## Now we get the coordinates related to each postal code

In [86]:
file = '/Users/diegop/Desktop/IBM_Certification/Course9.Applied_Data_Science_Capstone/Final_Assignments/Geospatial_Coordinates.csv'

loc = pd.read_csv(file)

loc.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [89]:
loc.shape

(103, 3)

In [90]:
#The order of the first dataframe matches the 2nd (in both position and shape)
df = df.sort_values(by ='Postal Code' )

In [94]:
df.head()

Postal Code      Borough                            Neighborhood
9          M1B  Scarborough                          Malvern, Rouge
18         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
27         M1E  Scarborough       Guildwood, Morningside, West Hill
36         M1G  Scarborough                                  Woburn
45         M1H  Scarborough                               Cedarbrae

In [100]:
# Reseting index for first dataframe

df = df.reset_index(drop=True)
df.head()

Postal Code      Borough                            Neighborhood
0         M1B  Scarborough                          Malvern, Rouge
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2         M1E  Scarborough       Guildwood, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                               Cedarbrae

In [108]:
# Merging both dataframes

df_merge = pd.merge(df,loc, on = 'Postal Code')
df_merge.head(10)

Postal Code      Borough                                     Neighborhood  \
0         M1B  Scarborough                                   Malvern, Rouge   
1         M1C  Scarborough           Rouge Hill, Port Union, Highland Creek   
2         M1E  Scarborough                Guildwood, Morningside, West Hill   
3         M1G  Scarborough                                           Woburn   
4         M1H  Scarborough                                        Cedarbrae   
5         M1J  Scarborough                              Scarborough Village   
6         M1K  Scarborough      Kennedy Park, Ionview, East Birchmount Park   
7         M1L  Scarborough                  Golden Mile, Clairlea, Oakridge   
8         M1M  Scarborough  Cliffside, Cliffcrest, Scarborough Village West   
9         M1N  Scarborough                      Birch Cliff, Cliffside West   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476  
5  43.744734 -79.239476  
6  43.727929 -79.262029  
7  43.711112 -79.284577  
8  43.716316 -79.239476  
9  43.692657 -79.264848

## The merged dataframe has 103 postal codes and 5 columns (Long and Lat were added)

In [109]:
df_merge.shape

(103, 5)

# Exploring and clustering

In [111]:
# Importing the necessary libraries

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [112]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(df_merge['Borough'].unique()),
        df_merge.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


In [113]:
# We now get Toronto coordinates

address = 'Toronto, TO'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))


The geograpical coordinate of Toronto City are 43.6534817, -79.3839347.


In [114]:
# Create a map of Toronto with neighborhoods superimposed on top.

map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_merge['Latitude'], df_merge['Longitude'], df_merge['Borough'], df_merge['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto


### Segmentarion using Foursquare

In [115]:
# Retrieving credentials
CLIENT_ID = 'BXI21QFED1RKO0G3LHN02A5SCGQIYF2CKSTO11QZN5AXPHJD' # your Foursquare ID
CLIENT_SECRET = 'LTIODAHZIH4R141FNA2FSSRPHG3AQFRUFFJ3HDPC1RLPUOTV' # your Foursquare Secret
VERSION = '20200520'
LIMIT = 30
print('My credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

My credentails:
CLIENT_ID: BXI21QFED1RKO0G3LHN02A5SCGQIYF2CKSTO11QZN5AXPHJD
CLIENT_SECRET:LTIODAHZIH4R141FNA2FSSRPHG3AQFRUFFJ3HDPC1RLPUOTV


In [123]:
# Name of the 3rd neighborhood and its borough.
print('The name of the 3rd Neighborhood is', df_merge.loc[2, 'Neighborhood'], 'and its borough is', df_merge.loc[2, 'Borough'])

The name of the 3rd Neighborhood is Guildwood, Morningside, West Hill and its borough is Scarborough


In [126]:
# We will focused on Scarborough
scar_data = df_merge[df_merge['Borough'] == 'Scarborough'].reset_index(drop=True)
scar_data.head(15)

Postal Code      Borough  \
0          M1B  Scarborough   
1          M1C  Scarborough   
2          M1E  Scarborough   
3          M1G  Scarborough   
4          M1H  Scarborough   
5          M1J  Scarborough   
6          M1K  Scarborough   
7          M1L  Scarborough   
8          M1M  Scarborough   
9          M1N  Scarborough   
10         M1P  Scarborough   
11         M1R  Scarborough   
12         M1S  Scarborough   
13         M1T  Scarborough   
14         M1V  Scarborough   

                                         Neighborhood   Latitude  Longitude  
0                                      Malvern, Rouge  43.806686 -79.194353  
1              Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497  
2                   Guildwood, Morningside, West Hill  43.763573 -79.188711  
3                                              Woburn  43.770992 -79.216917  
4                                           Cedarbrae  43.773136 -79.239476  
5                                 Scarborough Village  43.744734 -79.239476  
6         Kennedy Park, Ionview, East Birchmount Park  43.727929 -79.262029  
7                     Golden Mile, Clairlea, Oakridge  43.711112 -79.284577  
8     Cliffside, Cliffcrest, Scarborough Village West  43.716316 -79.239476  
9                         Birch Cliff, Cliffside West  43.692657 -79.264848  
10  Dorset Park, Wexford Heights, Scarborough Town...  43.757410 -79.273304  
11                                  Wexford, Maryvale  43.750072 -79.295849  
12                                          Agincourt  43.794200 -79.262029  
13            Clarks Corners, Tam O'Shanter, Sullivan  43.781638 -79.304302  
14  Milliken, Agincourt North, Steeles East, L'Amo...  43.815252 -79.284577

In [127]:
# Now we retrieve Scarborough coordinates from GEO
address = 'Scarborough, TO'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Scarborough are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Scarborough are 43.773077, -79.257774.


In [129]:
# Now we retrieve the corresponding neighborhood coordinates from the dataframe

neigh_lat = df_merge.loc[2, 'Latitude'] # neighborhood latitude value
neigh_long = df_merge.loc[2, 'Longitude'] # neighborhood longitude value

neigh_name = df_merge.loc[2, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neigh_name, 
                                                               neigh_lat, 
                                                               neigh_long))

Latitude and longitude values of Guildwood, Morningside, West Hill are 43.7635726, -79.1887115.


In [130]:
#### Now, let's get the top 100 venues that are in that neighborhood within a radius of 500 meters.

LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neigh_lat, 
    neigh_long, 
    radius, 
    LIMIT)

results = requests.get(url).json()
results


{'meta': {'code': 200, 'requestId': '5ec3ea811e152c001b24fa60'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Scarborough Village',
  'headerFullLocation': 'Scarborough Village, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 7,
  'suggestedBounds': {'ne': {'lat': 43.768072604500006,
    'lng': -79.18249216787879},
   'sw': {'lat': 43.7590725955, 'lng': -79.1949308321212}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4beee041e24d20a1cd857314',
       'name': 'RBC Royal Bank',
       'location': {'address': '4374 KINGSTON RD',
        'crossStreet': 'Kingston & Lawrence',
        'lat': 43.76678992471017,
        'lng': -79.19115118872593,
        '

In [131]:
# function that extracts the category of the venue

def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [134]:
# Cleaning the file and structure it as DF.
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head(7)

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:4: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  after removing the cwd from sys.path.


name           categories        lat        lng
0                RBC Royal Bank                 Bank  43.766790 -79.191151
1             G & G Electronics    Electronics Store  43.765309 -79.191537
2              Big Bite Burrito   Mexican Restaurant  43.766299 -79.190720
3         Enterprise Rent-A-Car  Rental Car Location  43.764076 -79.193406
4         Woburn Medical Centre       Medical Center  43.766631 -79.192286
5  Lawrence Ave E & Kingston Rd         Intersection  43.767704 -79.189490
6                      Eggsmart       Breakfast Spot  43.767800 -79.190466

In [133]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

7 venues were returned by Foursquare.


### Exploring Neighborhoods in Scarborough

In [259]:
# Function to retrieve venues from Scarborough
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [260]:
# Now we apply the function to the desired borough

scar_venues = getNearbyVenues(names=scar_data['Neighborhood'],
                                   latitudes=scar_data['Latitude'],
                                   longitudes=scar_data['Longitude']
                                  )

print(scar_venues.shape)


Malvern, Rouge
Rouge Hill, Port Union, Highland Creek
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
Kennedy Park, Ionview, East Birchmount Park
Golden Mile, Clairlea, Oakridge
Cliffside, Cliffcrest, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Wexford Heights, Scarborough Town Centre
Wexford, Maryvale
Agincourt
Clarks Corners, Tam O'Shanter, Sullivan
Milliken, Agincourt North, Steeles East, L'Amoreaux East
Steeles West, L'Amoreaux West
Upper Rouge
(92, 7)


In [261]:
# Checking the new dataframe 

scar_venues.head()

Neighborhood  Neighborhood Latitude  \
0                          Malvern, Rouge              43.806686   
1  Rouge Hill, Port Union, Highland Creek              43.784535   
2  Rouge Hill, Port Union, Highland Creek              43.784535   
3       Guildwood, Morningside, West Hill              43.763573   
4       Guildwood, Morningside, West Hill              43.763573   

   Neighborhood Longitude                      Venue  Venue Latitude  \
0              -79.194353                    Wendy’s       43.807448   
1              -79.160497      Royal Canadian Legion       43.782533   
2              -79.160497  Affordable Toronto Movers       43.787919   
3              -79.188711             RBC Royal Bank       43.766790   
4              -79.188711          G & G Electronics       43.765309   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.163085                   Bar  
2       -79.162977         Moving Target  
3       -79.191151                  Bank  
4       -79.191537     Electronics Store

In [262]:
# Unique categories

print('There are {} uniques categories.'.format(len(scar_venues['Venue Category'].unique())))

There are 55 uniques categories.


### Analizing each neighborhood


In [263]:
# one hot encoding
scar_onehot = pd.get_dummies(scar_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
scar_onehot['Neighborhood'] = scar_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [scar_onehot.columns[-1]] + list(scar_onehot.columns[:-1])
scar_onehot = scar_onehot[fixed_columns]



In [264]:
# Group

scar_grouped = scar_onehot.groupby('Neighborhood').mean().reset_index()
scar_grouped.shape

(16, 56)

In [265]:
# Let's print each neighborhood along with the top 5 most common venues
num_top_venues = 5

for hood in scar_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = scar_grouped[scar_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0  Latin American Restaurant   0.2
1                     Lounge   0.2
2             Breakfast Spot   0.2
3               Skating Rink   0.2
4             Clothing Store   0.2


----Birch Cliff, Cliffside West----
                   venue  freq
0  General Entertainment  0.25
1           Skating Rink  0.25
2                   Café  0.25
3        College Stadium  0.25
4    American Restaurant  0.00


----Cedarbrae----
                 venue  freq
0               Bakery  0.11
1                 Bank  0.11
2               Lounge  0.11
3      Thai Restaurant  0.11
4  Fried Chicken Joint  0.11


----Clarks Corners, Tam O'Shanter, Sullivan----
                venue  freq
0         Pizza Place  0.14
1            Pharmacy  0.14
2  Chinese Restaurant  0.07
3         Gas Station  0.07
4        Noodle House  0.07


----Cliffside, Cliffcrest, Scarborough Village West----
                 venue  freq
0  American Restaurant   0.5
1                Mot

In [266]:
# Function to retrieve most common venues
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [267]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = scar_grouped['Neighborhood']

for ind in np.arange(scar_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(scar_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt          Skating Rink   
1                      Birch Cliff, Cliffside West       College Stadium   
2                                        Cedarbrae                Lounge   
3          Clarks Corners, Tam O'Shanter, Sullivan              Pharmacy   
4  Cliffside, Cliffcrest, Scarborough Village West   American Restaurant   

   2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0         Breakfast Spot  Latin American Restaurant                Lounge   
1  General Entertainment               Skating Rink                  Café   
2        Thai Restaurant         Athletics & Sports                Bakery   
3            Pizza Place         Chinese Restaurant  Fast Food Restaurant   
4                  Motel                 Hobby Shop         Grocery Store   

   5th Most Common Venue  6th Most Common Venue   7th Most Common Venue  \
0         Clothing Store  Vietnamese Restaurant       Convenience Store   
1  Vietnamese Restaurant          Grocery Store             Gas Station   
2                   Bank            Gas Station        Hakka Restaurant   
3      Convenience Store           Noodle House      Italian Restaurant   
4  General Entertainment            Gas Station  Furniture / Home Store   

    8th Most Common Venue 9th Most Common Venue  10th Most Common Venue  
0   General Entertainment           Gas Station  Furniture / Home Store  
1  Furniture / Home Store   Fried Chicken Joint    Fast Food Restaurant  
2     Fried Chicken Joint  Caribbean Restaurant   Vietnamese Restaurant  
3             Coffee Shop   Fried Chicken Joint                    Bank  
4     Fried Chicken Joint  Fast Food Restaurant       Electronics Store

## Cluster Neighborhoods

In [268]:
# Run *k*-means to cluster the neighborhood into 3 clusters.
# set number of clusters
kclusters = 3

scar_grouped_clustering = scar_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(scar_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 1], dtype=int32)

In [269]:
# add clustering labels

neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

scar_merged = scar_data

# merge scar_grouped with scar_data to add latitude/longitude for each neighborhood
scar_merged = scar_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')


In [270]:
scar_merged.head(17) 

Postal Code      Borough  \
0          M1B  Scarborough   
1          M1C  Scarborough   
2          M1E  Scarborough   
3          M1G  Scarborough   
4          M1H  Scarborough   
5          M1J  Scarborough   
6          M1K  Scarborough   
7          M1L  Scarborough   
8          M1M  Scarborough   
9          M1N  Scarborough   
10         M1P  Scarborough   
11         M1R  Scarborough   
12         M1S  Scarborough   
13         M1T  Scarborough   
14         M1V  Scarborough   
15         M1W  Scarborough   
16         M1X  Scarborough   

                                         Neighborhood   Latitude  Longitude  \
0                                      Malvern, Rouge  43.806686 -79.194353   
1              Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497   
2                   Guildwood, Morningside, West Hill  43.763573 -79.188711   
3                                              Woburn  43.770992 -79.216917   
4                                           Cedarbrae  43.773136 -79.239476   
5                                 Scarborough Village  43.744734 -79.239476   
6         Kennedy Park, Ionview, East Birchmount Park  43.727929 -79.262029   
7                     Golden Mile, Clairlea, Oakridge  43.711112 -79.284577   
8     Cliffside, Cliffcrest, Scarborough Village West  43.716316 -79.239476   
9                         Birch Cliff, Cliffside West  43.692657 -79.264848   
10  Dorset Park, Wexford Heights, Scarborough Town...  43.757410 -79.273304   
11                                  Wexford, Maryvale  43.750072 -79.295849   
12                                          Agincourt  43.794200 -79.262029   
13            Clarks Corners, Tam O'Shanter, Sullivan  43.781638 -79.304302   
14  Milliken, Agincourt North, Steeles East, L'Amo...  43.815252 -79.284577   
15                      Steeles West, L'Amoreaux West  43.799525 -79.318389   
16                                        Upper Rouge  43.836125 -79.205636   

    Cluster Labels 1st Most Common Venue  2nd Most Common Venue  \
0              1.0  Fast Food Restaurant  Vietnamese Restaurant   
1              0.0         Moving Target                    Bar   
2              0.0                  Bank           Intersection   
3              0.0           Coffee Shop      Korean Restaurant   
4              0.0                Lounge        Thai Restaurant   
5              2.0                   Spa             Playground   
6              0.0           Coffee Shop      Convenience Store   
7              0.0              Bus Line                 Bakery   
8              0.0   American Restaurant                  Motel   
9              0.0       College Stadium  General Entertainment   
10             0.0     Indian Restaurant  Vietnamese Restaurant   
11             0.0                Bakery             Smoke Shop   
12             0.0          Skating Rink         Breakfast Spot   
13             0.0              Pharmacy            Pizza Place   
14             2.0            Playground                   Park   
15             0.0    Chinese Restaurant   Fast Food Restaurant   
16             NaN                   NaN                    NaN   

        3rd Most Common Venue      4th Most Common Venue  \
0             College Stadium              Grocery Store   
1       Vietnamese Restaurant              Grocery Store   
2              Breakfast Spot        Rental Car Location   
3       Vietnamese Restaurant          Convenience Store   
4          Athletics & Sports                     Bakery   
5       Vietnamese Restaurant                Coffee Shop   
6                 Bus Station             Discount Store   
7                Intersection              Metro Station   
8                  Hobby Shop              Grocery Store   
9                Skating Rink                       Café   
10     Thrift / Vintage Store         Chinese Restaurant   
11             Breakfast Spot  Middle Eastern Restaurant   
12  Latin American Restaurant     

In [271]:
# The last entry gave nothing but NaN, so it will be removed
scar_merged.drop(index=16, inplace=True)

In [272]:
scar_merged.tail(16)

Postal Code      Borough  \
0          M1B  Scarborough   
1          M1C  Scarborough   
2          M1E  Scarborough   
3          M1G  Scarborough   
4          M1H  Scarborough   
5          M1J  Scarborough   
6          M1K  Scarborough   
7          M1L  Scarborough   
8          M1M  Scarborough   
9          M1N  Scarborough   
10         M1P  Scarborough   
11         M1R  Scarborough   
12         M1S  Scarborough   
13         M1T  Scarborough   
14         M1V  Scarborough   
15         M1W  Scarborough   

                                         Neighborhood   Latitude  Longitude  \
0                                      Malvern, Rouge  43.806686 -79.194353   
1              Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497   
2                   Guildwood, Morningside, West Hill  43.763573 -79.188711   
3                                              Woburn  43.770992 -79.216917   
4                                           Cedarbrae  43.773136 -79.239476   
5                                 Scarborough Village  43.744734 -79.239476   
6         Kennedy Park, Ionview, East Birchmount Park  43.727929 -79.262029   
7                     Golden Mile, Clairlea, Oakridge  43.711112 -79.284577   
8     Cliffside, Cliffcrest, Scarborough Village West  43.716316 -79.239476   
9                         Birch Cliff, Cliffside West  43.692657 -79.264848   
10  Dorset Park, Wexford Heights, Scarborough Town...  43.757410 -79.273304   
11                                  Wexford, Maryvale  43.750072 -79.295849   
12                                          Agincourt  43.794200 -79.262029   
13            Clarks Corners, Tam O'Shanter, Sullivan  43.781638 -79.304302   
14  Milliken, Agincourt North, Steeles East, L'Amo...  43.815252 -79.284577   
15                      Steeles West, L'Amoreaux West  43.799525 -79.318389   

    Cluster Labels 1st Most Common Venue  2nd Most Common Venue  \
0              1.0  Fast Food Restaurant  Vietnamese Restaurant   
1              0.0         Moving Target                    Bar   
2              0.0                  Bank           Intersection   
3              0.0           Coffee Shop      Korean Restaurant   
4              0.0                Lounge        Thai Restaurant   
5              2.0                   Spa             Playground   
6              0.0           Coffee Shop      Convenience Store   
7              0.0              Bus Line                 Bakery   
8              0.0   American Restaurant                  Motel   
9              0.0       College Stadium  General Entertainment   
10             0.0     Indian Restaurant  Vietnamese Restaurant   
11             0.0                Bakery             Smoke Shop   
12             0.0          Skating Rink         Breakfast Spot   
13             0.0              Pharmacy            Pizza Place   
14             2.0            Playground                   Park   
15             0.0    Chinese Restaurant   Fast Food Restaurant   

        3rd Most Common Venue      4th Most Common Venue  \
0             College Stadium              Grocery Store   
1       Vietnamese Restaurant              Grocery Store   
2              Breakfast Spot        Rental Car Location   
3       Vietnamese Restaurant          Convenience Store   
4          Athletics & Sports                     Bakery   
5       Vietnamese Restaurant                Coffee Shop   
6                 Bus Station             Discount Store   
7                Intersection              Metro Station   
8                  Hobby Shop              Grocery Store   
9                Skating Rink                       Café   
10     Thrift / Vintage Store         Chinese Restaurant   
11             Breakfast Spot  Middle Eastern Restaurant   
12  Latin American Restaurant                     Lounge   
13         Chinese Restaurant       Fast Food Restaurant   
14      Vietnamese Restaurant                Coffee Shop   
15     Thrift / Vintage Store  

In [273]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(scar_merged['Latitude'], scar_merged['Longitude'], scar_merged['Neighborhood'], scar_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color = rainbow[int(cluster)-1],
        #color=rainbow[cluster-1],
        fill=True,
        #fill_color=rainbow[cluster-1],
        fill_color=rainbow[int(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Finally, we examine each cluster

In [276]:
# First cluster
scar_merged.loc[scar_merged['Cluster Labels'] == 0, scar_merged.columns[[1] + list(range(5, scar_merged.shape[1]))]]



Borough  Cluster Labels 1st Most Common Venue  2nd Most Common Venue  \
1   Scarborough             0.0         Moving Target                    Bar   
2   Scarborough             0.0                  Bank           Intersection   
3   Scarborough             0.0           Coffee Shop      Korean Restaurant   
4   Scarborough             0.0                Lounge        Thai Restaurant   
6   Scarborough             0.0           Coffee Shop      Convenience Store   
7   Scarborough             0.0              Bus Line                 Bakery   
8   Scarborough             0.0   American Restaurant                  Motel   
9   Scarborough             0.0       College Stadium  General Entertainment   
10  Scarborough             0.0     Indian Restaurant  Vietnamese Restaurant   
11  Scarborough             0.0                Bakery             Smoke Shop   
12  Scarborough             0.0          Skating Rink         Breakfast Spot   
13  Scarborough             0.0              Pharmacy            Pizza Place   
15  Scarborough             0.0    Chinese Restaurant   Fast Food Restaurant   

        3rd Most Common Venue      4th Most Common Venue  \
1       Vietnamese Restaurant              Grocery Store   
2              Breakfast Spot        Rental Car Location   
3       Vietnamese Restaurant          Convenience Store   
4          Athletics & Sports                     Bakery   
6                 Bus Station             Discount Store   
7                Intersection              Metro Station   
8                  Hobby Shop              Grocery Store   
9                Skating Rink                       Café   
10     Thrift / Vintage Store         Chinese Restaurant   
11             Breakfast Spot  Middle Eastern Restaurant   
12  Latin American Restaurant                     Lounge   
13         Chinese Restaurant       Fast Food Restaurant   
15     Thrift / Vintage Store                       Bank   

     5th Most Common Venue  6th Most Common Venue   7th Most Common Venue  \
1    General Entertainment            Gas Station  Furniture / Home Store   
2        Electronics Store         Medical Center      Mexican Restaurant   
3            Grocery Store  General Entertainment             Gas Station   
4                     Bank            Gas Station        Hakka Restaurant   
6         Department Store             Hobby Shop           Grocery Store   
7                     Park            Bus Station          Ice Cream Shop   
8    General Entertainment            Gas Station  Furniture / Home Store   
9    Vietnamese Restaurant          Grocery Store             Gas Station   
10               Pet Store      Convenience Store   General Entertainment   
11   Vietnamese Restaurant      Convenience Store   General Entertainment   
12          Clothing Store  Vietnamese Restaurant       Convenience Store   
13       Convenience Store           Noodle House      Italian Restaurant   
15  Furniture / Home Store         Breakfast Spot          Sandwich Place   

     8th Most Common Venue   9th Most Common Venue  10th Most Common Venue  
1      Fried Chicken Joint    Fast Food Restaurant       Electronics Store  
2    Vietnamese Restaurant       Convenience Store             Gas Station  
3   Furniture / Home Store     Fried Chicken Joint    Fast Food Restaurant  
4      Fried Chicken Joint    Caribbean Restaurant   Vietnamese Restaurant  
6    General Entertainment             Gas Station  Furniture / Home Store  
7             Soccer Field     Fried Chicken Joint       Convenience Store  
8      Fried Chicken Joint    Fast Food Restaurant       Electronics Store  
9   Furniture / Home Store     Fried Chicken Joint    Fast Food Restaurant  
10             Gas Station  Furniture / Home Store     Fried Chicken Joint  
11             Gas Station  Furniture / Home Store     Fried Chicken Joint  
12   General Entertainment             Gas Station  Furniture / Home Store  
13             Coffee Shop     Fried Chicke

In [277]:
# Second cluster
scar_merged.loc[scar_merged['Cluster Labels'] == 1, scar_merged.columns[[1] + list(range(5, scar_merged.shape[1]))]]


Borough  Cluster Labels 1st Most Common Venue  2nd Most Common Venue  \
0  Scarborough             1.0  Fast Food Restaurant  Vietnamese Restaurant   

  3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
0       College Stadium         Grocery Store  General Entertainment   

  6th Most Common Venue   7th Most Common Venue 8th Most Common Venue  \
0           Gas Station  Furniture / Home Store   Fried Chicken Joint   

  9th Most Common Venue 10th Most Common Venue  
0     Electronics Store         Discount Store

In [278]:
# Third cluster
scar_merged.loc[scar_merged['Cluster Labels'] == 2, scar_merged.columns[[1] + list(range(5, scar_merged.shape[1]))]]


Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
5   Scarborough             2.0                   Spa            Playground   
14  Scarborough             2.0            Playground                  Park   

    3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
5   Vietnamese Restaurant           Coffee Shop  General Entertainment   
14  Vietnamese Restaurant           Coffee Shop  General Entertainment   

   6th Most Common Venue   7th Most Common Venue 8th Most Common Venue  \
5            Gas Station  Furniture / Home Store   Fried Chicken Joint   
14           Gas Station  Furniture / Home Store   Fried Chicken Joint   

   9th Most Common Venue 10th Most Common Venue  
5   Fast Food Restaurant      Electronics Store  
14  Fast Food Restaurant      Electronics Store

#### Not much to add. I ran the code with different values of k (clusters). Cluster #1 has all the nice restaurants while cluster#2 has the only fast food restaurant of the whole borough.